## Topic: 

<p>Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with $1$ and $2$, the first $10$ terms will be:
$$1, 2, 3, 5, 8, 13, 21, 34, 55, 89, \dots$$</p>
<p>By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.</p>

In [33]:
# Brute force will work but i am trying to find a patter because we are in the one-minute rule programming

In [15]:
import time

def fibonacci_tabular(n):
    if n == 1:
        return 1
    elif n == 2:
        return 2
    
    # Create a table to store Fibonacci values
    fib = [0] * (n + 1)
    fib[1] = 1
    fib[2] = 2
    
    # Fill the table in a bottom-up manner
    for i in range(3, n + 1):
        fib[i] = fib[i - 1] + fib[i - 2]
    
    return fib[n]

# Example usage:
n = 100

# Start time measurement
start_time = time.time()

result = fibonacci_tabular(n)

# End time measurement
end_time = time.time()

print(f"Fibonacci({n}) =", result)
print("Execution time:", end_time - start_time, "seconds")

Fibonacci(100) = 573147844013817084101
Execution time: 4.315376281738281e-05 seconds


In [3]:
# Define a dictionary to store already computed Fibonacci values
memo = {}

def fibonacci(n):
    # Base cases
    if n == 1 :
        return 1
    elif n==2: 
        return 2
    
    # Check if result is already computed
    if n not in memo:
        # Recursive calculation with memoization
        memo[n] = fibonacci(n - 1) + fibonacci(n - 2)
    
    return memo[n]

# Example usage:
n = 1000

# Start time measurement
start_time = time.time()

result = fibonacci(n)

# End time measurement
end_time = time.time()

print(f"Fibonacci({n}) =", result)
print("Execution time:", end_time - start_time, "seconds")

Fibonacci(1000) = 70330367711422815821835254877183549770181269836358732742604905087154537118196933579742249494562611733487750449241765991088186363265450223647106012053374121273867339111198139373125598767690091902245245323403501
Execution time: 0.0026221275329589844 seconds


In [26]:
# Brute force solution using fibo tabular 

# Start time measurement
start_time = time.time()

res_fibo = 0
index_fibo = 1
ans = 0
barrier = 4_000_000

while (res_fibo < barrier):
    index_fibo +=1
    res_fibo = fibonacci_tabular(index_fibo)
    print(index_fibo, res_fibo)
    if res_fibo % 2 == 0 and res_fibo < barrier:
        print("value use in the sum", res_fibo)
        ans += res_fibo
print("Barrier break at Fibo indice: ", index_fibo, "value: ", res_fibo, "\n")

# End time measurement
end_time = time.time()

print(f"Index before it breaks barrier =", index_fibo - 1)
print(f"Sum even even-value term of Fibonacci =", ans)
print("Execution time:", end_time - start_time, "seconds")

2 2
value use in the sum 2
3 3
4 5
5 8
value use in the sum 8
6 13
7 21
8 34
value use in the sum 34
9 55
10 89
11 144
value use in the sum 144
12 233
13 377
14 610
value use in the sum 610
15 987
16 1597
17 2584
value use in the sum 2584
18 4181
19 6765
20 10946
value use in the sum 10946
21 17711
22 28657
23 46368
value use in the sum 46368
24 75025
25 121393
26 196418
value use in the sum 196418
27 317811
28 514229
29 832040
value use in the sum 832040
30 1346269
31 2178309
32 3524578
value use in the sum 3524578
33 5702887
Barrier break at Fibo indice:  33 value:  5702887 

Index before it breaks barrier = 32
Sum even even-value term of Fibonacci = 4613732
Execution time: 0.001641988754272461 seconds


**Naive solution:**
A direct translation of the problem statement would be a program like this

In [28]:
# Official solution 1
# Start time measurement
start_time = time.time()

# official solution - no need to compute no save every fibo
limit = 4_000_000
ans = 0
a = 1
b = 1

while b < limit: 
    if b%2 == 0:
        ans +=b
    h = a + b
    a = b
    b = h
print()

# End time measurement
end_time = time.time()

# print(f"Index before it breaks barrier =", index_fibo - 1)
print(f"Sum even even-value term of Fibonacci =", ans)
print("Execution time:", end_time - start_time, "seconds")


Sum even even-value term of Fibonacci = 4613732
Execution time: 0.0011761188507080078 seconds


**Now let try to improve solution:**

Now let us see if we can get rid of the testing for even values.
Here is the beginning of the Fibonacci sequence with even numbers in red:
- 1 1 2 3 5 8 13 21 34 55 89 144 ...
- a b c a b c a b c a b c

It is easy to prove that every third Fibonacci number is even.
It is not so difficult to change the program somewhat so that only every third number is
added:

In [31]:
# Official solution 2


# Start time measurement
start_time = time.time()

# official solution - no need to compute no save every fibo
limit = 4_000_000
ans = 0
a = 1
b = 1
c = a + b

while c < limit: 
    ans = ans + c
    a = b + c
    b = c + a
    c = a + b

# End time measurement
end_time = time.time()

# print(f"Index before it breaks barrier =", index_fibo - 1)
print(f"Sum even even-value term of Fibonacci =", ans)
print("Execution time:", end_time - start_time, "seconds")

Sum even even-value term of Fibonacci = 4613732
Execution time: 0.0003058910369873047 seconds


**Now let try to improve solution:**

There is another beautiful structure hidden beneath this problem:
If we only write the even numbers:
- 2 8 34 144...

it seems that they obey the following recursive relation: 
- E(n)=4*E(n-1)+E(n-2).

If we can prove that for the Fibonacci numbers the formula 
- F(n)=4*F(n-3)+F(n-6) 

holds we have proven this recursion.

In [32]:
# Official solution 3


# Start time measurement
start_time = time.time()

# Starting values for the first two even Fibonacci numbers
limit = 4_000_000
e1, e2 = 2, 8
total_sum = e1 + e2

# Generate subsequent even Fibonacci numbers using the recurrence relation
while True: 
    # Calculate the next even Fibonacci number
    e3 = 4 * e2 + e1
    if e3 > limit:
        break
        
    # Add to the sum and update variables for the next iteration
    total_sum += e3
    e1, e2 = e2, e3

# End time measurement
end_time = time.time()

# print(f"Index before it breaks barrier =", index_fibo - 1)
print(f"Sum even even-value term of Fibonacci =", total_sum)
print("Execution time:", end_time - start_time, "seconds")

Sum even even-value term of Fibonacci = 4613732
Execution time: 0.0003180503845214844 seconds
